File: frontend/src/api/course.ts
import { request } from "./request"

export interface Course {
  id: number
  name: string
  teacher: string
  school: string
  department: string
  created_at: string
}

export interface CourseRating {
  id: number
  course_id: number
  user_id: number
  teaching_rating: number
  difficulty: number
  grading: number
  comment: string
  is_anonymous: boolean
  helpful_count: number
  created_at: string
  nickname?: string
  avatar?: string
}

export interface CourseDetail {
  id: number
  name: string
  teacher: string
  school: string
  department: string
  avg_teaching: number
  avg_difficulty: number
  avg_grading: number
  rating_count: number
  user_rating?: CourseRating | null
  created_at: string
}

export function searchCourses(q: string) {
  return request<Course[]>({ url: "/courses/search", method: "GET", data: { q } as any })
}

export function getCourseDetail(id: number) {
  return request<CourseDetail>({ url: `/courses/${id}`, method: "GET" })
}

export function createRating(courseId: number, data: {
  teaching_rating: number
  difficulty: number
  grading: number
  comment?: string
  is_anonymous?: boolean
}) {
  return request<{ id: number }>({ url: `/courses/${courseId}/rating`, method: "POST", data })
}

export function listRatings(courseId: number, params?: { order?: string; offset?: number; limit?: number }) {
  return request<CourseRating[]>({ url: `/courses/${courseId}/ratings`, method: "GET", data: params as any })
}

export function getUserRating(courseId: number) {
  return request<CourseRating | null>({ url: `/courses/${courseId}/my-rating`, method: "GET" })
}

export function markHelpful(ratingId: number) {
  return request<{ message: string }>({ url: `/courses/rating/${ratingId}/helpful`, method: "POST" })
}

File: backend/internal/model/course.go
package model

import "time"

type Course struct {
	ID         int64     `json:"id"`
	Name       string    `json:"name"`
	Teacher    string    `json:"teacher"`
	School     string    `json:"school"`
	Department string    `json:"department"`
	CreatedAt  time.Time `json:"created_at"`
}

type CourseRating struct {
	ID             int64     `json:"id"`
	CourseID       int64     `json:"course_id"`
	UserID         int64     `json:"user_id"`
	TeachingRating int       `json:"teaching_rating"`
	Difficulty     int       `json:"difficulty"`
	Grading        int       `json:"grading"`
	Comment        string    `json:"comment"`
	IsAnonymous    bool      `json:"is_anonymous"`
	HelpfulCount   int       `json:"helpful_count"`
	Status         int       `json:"status"`
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
	Nickname       string    `json:"nickname,omitempty"`
	Avatar         string    `json:"avatar,omitempty"`
}

type CreateRatingRequest struct {
	TeachingRating int    `json:"teaching_rating" binding:"required,min=1,max=5"`
	Difficulty     int    `json:"difficulty" binding:"required,min=1,max=5"`
	Grading        int    `json:"grading" binding:"required,min=1,max=5"`
	Comment        string `json:"comment"`
	IsAnonymous    bool   `json:"is_anonymous"`
}

type CourseDetail struct {
	Course
	AvgTeaching float64        `json:"avg_teaching"`
	AvgDifficulty float64      `json:"avg_difficulty"`
	AvgGrading   float64       `json:"avg_grading"`
	RatingCount  int           `json:"rating_count"`
	UserRating   *CourseRating `json:"user_rating,omitempty"`
}
